In [45]:
import pandas as pd
from myst_nb import glue
import pydicom as dc

In [46]:
import uuid
from datetime import datetime
import pytz

In [47]:
estructura=dc.read_file('Ficheros/EstructuraSimple.dcm')
plansimple=dc.read_file('Ficheros/PlanSimple.dcm')
plancomplejo2=dc.read_file('Ficheros/PlanComplejo2.dcm')

# Anexo 1: Creación y exportación Dicom

## Introducción

Como proyecto paralelo a la reconstrucción completamente automática y utilizando la herramienta de [anotación](capitulo:anotation) que hemos desarrollado, podemos utilizar el resultado que proporciona la herramienta para hacer un uso clínico del mismo, es decir, incorporar la reconstrucción realizada al flujo de trabajo. Para ello vamos a investigar la estrura de los ficheros dicom de imágenes y ejemplos de exportación del plan para generar uno que pueda ser importado en Oncentra con la reconstrucción ya hecha.

## Tags comunes entre imágenes y plan: 

Hay 23 tags comunes entre un fichero de secuencia de imágenes y uno tipo RTPlan. De ellos, 14 poseen el mismo valor y por lo tanto el mismo viene definido de la MRI. La lista es:

<ul><li>AccessionNumber</li><li>FrameOfReferenceUID</li><li>PatientBirthDate</li><li>PatientID</li><li>PatientName</li><li>PatientSex</li><li>PositionReferenceIndicator</li><li>ReferringPhysicianName</li><li>SpecificCharacterSet</li><li>StudyDate</li><li>StudyDescription</li><li>StudyID</li><li>StudyInstanceUID</li><li>StudyTime</li></ul>

### Tags de diferente valor

Los 9 que nos quedan se muestran en la tabla siguiente con valores ejemplo de tres planes diferentes y de un archivo de imagen:

In [48]:
campos=pd.read_csv('data/camposnocomunes.csv')
campos.set_index('tag')

Plan1  \
tag                                                                        
Manufacturer                                                   Nucletron   
ManufacturerModelName                                           Oncentra   
Modality                                                          RTPLAN   
SOPClassUID                                1.2.840.10008.5.1.4.1.1.481.5   
SOPInstanceUID         1.3.6.1.4.1.2452.6.2743304011.1151359869.12204...   
SeriesInstanceUID      1.3.6.1.4.1.2452.6.3396086542.1231529005.38488...   
SeriesNumber                                                           1   
SoftwareVersions                                           OTP V4.5.3.30   
StationName                                                   RT-BRACHYT   

                                                                   Plan2  \
tag                                                                        
Manufacturer                                                   Nucletron   
ManufacturerModelName                                           Oncentra   
Modality                                                          RTPLAN   
SOPClassUID                                1.2.840.10008.5.1.4.1.1.481.5   
SOPInstanceUID         1.3.6.1.4.1.2452.6.2774509762.1153355867.34264...   
SeriesInstanceUID      1.3.6.1.4.1.2452.6.2747787005.1263068138.27367...   
SeriesNumber                                                           1   
SoftwareVersions                                          OTP V4.5.2.152   
StationName                                                     PC009761   

                                                                   Plan3  \
tag                                                                        
Manufacturer                                                   Nucletron   
ManufacturerModelName                                           Oncentra   
Modality                                                          RTPLAN   
SOPClassUID                                1.2.840.10008.5.1.4.1.1.481.5   
SOPInstanceUID         1.3.6.1.4.1.2452.6.2633641423.1222660011.25835...   
SeriesInstanceUID      1.3.6.1.4.1.2452.6.1938074019.1188018291.41990...   
SeriesNumber                                                           1   
SoftwareVersions                                           OTP V4.5.3.30   
StationName                                                     PC009761   

                                                                  Imagen  
tag                                                                       
Manufacturer                                          GE MEDICAL SYSTEMS  
ManufacturerModelName                                         Signa HDxt  
Modality                                                              MR  
SOPClassUID                                    1.2.840.10008.5.1.4.1.1.4  
SOPInstanceUID         1.3.6.1.4.1.2452.6.1019122476.1241727211.17704...  
SeriesInstanceUID      1.3.6.1.4.1.2452.6.3150457222.1106452005.39453...  
SeriesNumber                                                           5  
SoftwareVersions       ['24', 'LX', 'MR Software release:HD16.0_V03_1...  
StationName                                                     NHFEMR04

Hay una serie de ellos que no nos van a suponer ningún problema. No obstante detengámonos en cada uno de ellos

#### Manufacter y ManufacturerModelName

Podemos poner en ambos el nombre que queramos por ejemplo **IRIMED** y **UtrechtTool**

#### Modality

Esta debe ser obligatoriamente **RTPlan**

#### SeriesNumber

En esta pondremos **1**

#### SoftwareVersion y StationName

De momento pondremos **1.0** para la primera y **3DSlicer** para la segunda.

Ahora entramos en los campos no tan evidentes

#### SOPClassUID

Por lo que vemos en los diferentes planes, este valor es común. EL valor **1.2.840.10008.5.1.4.1.1.481.5** corresponde a un _Radiation Therapy Plan Storage_.

#### SOPInstanceUID y SeriesInstanceUID

En el caso de estos dos valores las cosas no están tan claras. Parece que poseeen el mismo prefijo que sus equivalentes en el fichero de imagen. La segunda parte corresponde a un número aleatorio de entre 39 y 40 cifras separado en grupos de 10 por puntos. Así pues los generamos con el siguiente código:

In [49]:
prefix='1.3.6.1.4.1.2452.6.'
def UIDgenerator(prefix):
    while True:
        uid=str(uuid.uuid4().int)
        if (len(uid)>=39):
             return prefix+uid[0:10]+'.'+uid[10:20]+'.'+uid[20:30]+'.'+uid[30:]
UIDgenerator(prefix)

'1.3.6.1.4.1.2452.6.2615407535.8834935661.0130770635.823510730'

## Tags no comunes

### Tags no sequence

<ul><li>ApprovalStatus</li><li>BrachyTreatmentTechnique</li><li>BrachyTreatmentType</li><li>InstanceCreationDate</li><li>InstanceCreationTime</li><li>InstitutionalDepartmentName</li><li>OperatorsName</li><li>RTPlanDate</li><li>RTPlanGeometry</li><li>RTPlanLabel</li><li>RTPlanName</li><li>RTPlanTime</li><li>TimezoneOffsetFromUTC</li></ul>

#### ApprovalStatus

El valor será **UNAPPROVED**

#### BrachyTreatmentTechnique

Pondremos valor **INTERSTITIAL**

#### BrachyTreatmentType
El valor es **HDR**

#### InstanceCreationDate y InstanceCreationTime
Fecha del día de la creación del plan en formato AAAAMMDD y la hora HHMMSS

In [50]:
Hoy=datetime.today()
print(Hoy.strftime("%Y%m%d"),Hoy.strftime("%H%M%S"))

20210326 221338


#### OperatorsName

Cuanquiera. **Plan** por ejemplo.

#### RTPlanDate

La misma que **InstanceCreationTime**

#### RTPlanGeometry
El valor es **PATIENT**

#### RTPlanLabel y RTPlanName
Por ejemplo **IRIMED** en ambos

#### RTPlanTime
El mismo valor que InstanceCreationTime

#### TimezoneOffsetFromUTC
La función que nos da el valor del tag se muestra a continuación

In [51]:
def DiferenciaUTC(X='Europe/Madrid'):
    delta=pytz.timezone(X).utcoffset(datetime.today())
    delta=delta.days*24+delta.seconds//3600
    return "{:=+03d}00".format(delta)

In [52]:
DiferenciaUTC()

'+0100'

In [53]:
DiferenciaUTC('US/Pacific')

'-0700'

### Tags sequence

La lista de tags de secuencia son los siguientes

<ul><li>FractionGroupSequence</li><li>ReferencedStructureSetSequence</li><li>SourceSequence</li><li>TreatmentMachineSequence</li><li>ApplicationSetupSequence</li></ul>

Salvo que se diga lo contrario siempre nos referiremos a la posición 0 de la secuencia

#### FractionGroupSequence
Dejaremos el siguiente tag por defecto

In [54]:
print(plansimple.data_element('FractionGroupSequence').value[0])

(300a, 0071) Fraction Group Number               IS: "1"
(300a, 0078) Number of Fractions Planned         IS: "1"
(300a, 0080) Number of Beams                     IS: "0"
(300a, 00a0) Number of Brachy Application Setups IS: "1"
(300c, 000a)  Referenced Brachy Application Setup Sequence  1 item(s) ---- 
   (300c, 000c) Referenced Brachy Application Setup IS: "0"
   ---------


#### ReferencedStructureSetSequence
El aspecto del tag es el siguiente:

In [55]:
print(plansimple.data_element('ReferencedStructureSetSequence').value[0])

(0008, 1150) Referenced SOP Class UID            UI: RT Structure Set Storage
(0008, 1155) Referenced SOP Instance UID         UI: 1.3.6.1.4.1.2452.6.2664562426.1288265039.467152794.2028798075


Simplemente debemos localizar en el fichero de estructuras los campos SOPClassUID y SOPInstanceUID e incorporar esos valores a los sub tags. 

#### SourceSequence
Esta secuencia tiene el siguiente aspecto:

In [56]:
print(plansimple.data_element('SourceSequence').value[0])

(300a, 0212) Source Number                       IS: "0"
(300a, 0214) Source Type                         CS: 'LINE'
(300a, 0226) Source Isotope Name                 LO: 'Ir-192'
(300a, 0228) Source Isotope Half Life            DS: "73.81"
(300a, 0229) Source Strength Units               CS: 'AIR_KERMA_RATE'
(300a, 022a) Reference Air Kerma Rate            DS: "21733.8274231908"
(300a, 022c) Source Strength Reference Date      DA: '20210323'
(300a, 022e) Source Strength Reference Time      TM: '090355'
(300b, 0010) Private Creator                     LO: 'NUCLETRON'
(300b, 1006) Private tag data                    DS: "-100663.0"
(300b, 1007) Private tag data                    DS: "-100693.0"
(300b, 1008) Private tag data                    DT: '20201229143800'
(300b, 100c) Private tag data                    DS: None


Ignoraremos los private tag data. Para los campos **ReferenceAirKermaRate**,**SourceStrength ReferenceDate** y **SourceStrengthReferenceTime** tenemos dos opciones: O ponemos valores genéricos de fuente, lo más cómodo, o ponemos los valores de la funete que vayamos a utilizar. Lo demás se puede dejar como en el ejemplo 

#### ApplicationSetupSequence
En el ejemplo siguiente los dos primeros tags pueden quedar tal cual.

```
(300a, 0232) Application Setup Type              CS: 'GYNECOLOGY'
(300a, 0234) Application Setup Number            IS: "0"
(300a, 0250) Total Reference Air Kerma           DS: "3315.31795903434"
```

El tercero lo podemos calcular como el cociente entre el campo **ReferenceAirKermaRate** y el tiempo total de radiación expresado en horas que será la suma de los tiempos totales de cada fuente, en nuestro caso 1, y de cada uno de los canales.

Para el ejemplo que nos ocupa el plan se llama plancomplejo2 y los valores que buscamos son:

In [57]:
ReferenceAirKermaRate=plancomplejo2.data_element('SourceSequence').value[0].ReferenceAirKermaRate
Secuencia=plancomplejo2.data_element('ApplicationSetupSequence').value[0]
TiempoTotal=(Secuencia.ChannelSequence[0].ChannelTotalTime+Secuencia.ChannelSequence[1].ChannelTotalTime+Secuencia.ChannelSequence[2].ChannelTotalTime)/3600
TotalReferenceAirKerma=ReferenceAirKermaRate*TiempoTotal
print(TotalReferenceAirKerma)

3315.317959034321


Que como vemos coincide con lo que había en la cabecera. Con respecto al tiempo total, podemos elegir el tiempo que queramos por posición. Este cálculo lo hacemos para que simplemente sea consistente el fichero.
Por otra parte veamos ahora el aspecto de cada uno de los ChannelSequence

In [58]:
Secuencia.ChannelSequence[0]

(3006, 0084) Referenced ROI Number               IS: None
(300a, 0110) Number of Control Points            IS: "14"
(300a, 0282) Channel Number                      IS: "1"
(300a, 0284) Channel Length                      DS: "1500.0"
(300a, 0286) Channel Total Time                  DS: "82.6262964446422"
(300a, 0288) Source Movement Type                CS: 'STEPWISE'
(300a, 0290) Source Applicator Number            IS: "1"
(300a, 0291) Source Applicator ID                SH: '1'
(300a, 0292) Source Applicator Type              CS: 'FLEXIBLE'
(300a, 0294) Source Applicator Name              LO: ''
(300a, 0296) Source Applicator Length            DS: "37.0503443100255"
(300a, 02a0) Source Applicator Step Size         DS: "2.5"
(300a, 02a2) Transfer Tube Number                IS: "1"
(300a, 02a4) Transfer Tube Length                DS: None
(300a, 02c8) Final Cumulative Time Weight        DS: "6.33436977863312"
(300a, 02d0)  Brachy Control Point Sequence  14 item(s) ---- 
   (300a, 0112)

Ignoramos como siempre los campos privados. Del tiempo total ya hemos hablado así que lo saltaremos. El **SourceApplicatorLength** debe ser mayor que el recorrido total que hace la fuente a través de los canales. Parece que coincide con lo que reconstruimos sobre la imagen. Al **FinalCumulativeTimeWeight** le pasa lo mismo que al tiempo, mientras sea consistente no pasa nada. Vamos ahora al **BrachyControlPointSequence**.
Siempre tendrá un número par de elementos, veamos un ejemplo:

In [59]:
Secuencia.ChannelSequence[0].BrachyControlPointSequence[6]

(300a, 0112) Control Point Index                 IS: "6"
(300a, 02d2) Control Point Relative Position     DS: "7.5"
(300a, 02d4) Control Point 3D Position           DS: [-24.512237, -0.755723, 4.734024]
(300a, 02d6) Cumulative Time Weight              DS: "2.60867154598236"

In [60]:
Secuencia.ChannelSequence[0].BrachyControlPointSequence[7]

(300a, 0112) Control Point Index                 IS: "7"
(300a, 02d2) Control Point Relative Position     DS: "7.5"
(300a, 02d4) Control Point 3D Position           DS: [-24.512237, -0.755723, 4.734024]
(300a, 02d6) Cumulative Time Weight              DS: "3.57875949144363"

Los elementos van numerados de manera correlativa por un índice. La posición relativa es un múltiplo del paso de fuente, en este caso 2.5. El **ControlPoint3DPosition** nos vendrá del programa de reconstrucción descrito en el capítulo sobre [anotación](capitulo:anotation). El **CumulativeTimeWeight** al igual que el tiempo lo definiremos consistente con lo que elijamos para cada canal